File: src/main/java/zh/lingvo/domain/changepatterns/helpers/es/EsNounFormHelper.java
package zh.lingvo.domain.changepatterns.helpers.es;

import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSet;
import zh.lingvo.domain.LinguisticCategory;
import zh.lingvo.domain.changepatterns.helpers.WordFormsHelper;
import zh.lingvo.domain.words.Name;
import zh.lingvo.domain.words.Word;

import java.util.List;
import java.util.Map;
import java.util.Set;

import static zh.lingvo.domain.NounLinguisticCategories.PLURAL_NOMINATIVE;
import static zh.lingvo.domain.NounLinguisticCategories.SINGULAR_NOMINATIVE;

public class EsNounFormHelper implements WordFormsHelper {
    private static final Set<LinguisticCategory[]> forms = ImmutableSet.of(SINGULAR_NOMINATIVE, PLURAL_NOMINATIVE);
    private static final Set<String> VOWELS = ImmutableSet.of("e", "a", "o", "i", "u", "y",
            "é", "á", "ó", "í", "ú", "ü");

    @Override
    public Map<LinguisticCategory[], String> getForms(Word word, List<Name> formExceptions) {
        String baseForm = word.getName().getValue();
        return ImmutableMap.of(
                SINGULAR_NOMINATIVE, baseForm,
                PLURAL_NOMINATIVE, appendS(baseForm));
    }

    private String appendS(String word) {
        if (endsWithVowel(word))
            return substitute(word, 0, "s");
        if (endsWithIon(word))
            return substitute(word, 2, "ones");
        if (endsWithZ(word))
            return substitute(word, 1, "ces");
        return substitute(word, 0, "es");
    }

    private boolean endsWithVowel(String word) {
        return endingIsInSet(word, VOWELS);
    }

    private boolean endsWithIon(String word) {
        return word.endsWith("ción") || word.endsWith("sión");
    }

    private boolean endsWithZ(String word) {
        return word.endsWith("z");
    }

    @Override
    public Set<LinguisticCategory[]> getLanguageForms() {
        return EsNounFormHelper.forms;
    }
}

File: src/main/kotlin/zh/lingvo/domain/changepatterns/helpers/es/EsNounFormHelper.kt
package zh.lingvo.domain.changepatterns.helpers.es

import zh.lingvo.domain.changepatterns.BasicNounChangeModel
import zh.lingvo.domain.changepatterns.helpers.AbstractWordFormsHelper
import zh.lingvo.domain.forms.NounDeclensionsCategory.PLURAL_NOMINATIVE
import zh.lingvo.domain.forms.NounDeclensionsCategory.SINGULAR_NOMINATIVE
import zh.lingvo.domain.forms.WordForms
import zh.lingvo.domain.languages.Spanish
import zh.lingvo.domain.words.Name
import zh.lingvo.domain.words.Word

private val FORMS = setOf(SINGULAR_NOMINATIVE, PLURAL_NOMINATIVE)
private val BASE_VOWELS = setOf("e", "a", "o", "i", "u", "y")
private val VOWELS = setOf("e", "a", "o", "i", "u", "y", "é", "á", "ó", "í", "ú", "ü")
private val WEAK_CONSONANTS = setOf("s", "n")

private val STRESSED_VOWEL_PATTERN = ".*[éáóíú].*".toRegex().toPattern()

private val CHANGE_MODEL = BasicNounChangeModel(Spanish.getInstance())

class EsNounFormHelper : AbstractWordFormsHelper() {
    override fun getForms(word: Word, formExceptions: List<Name>): WordForms {
        val baseForm = word.name.value
        val wordForms = mapOf(
                SINGULAR_NOMINATIVE to baseForm,
                PLURAL_NOMINATIVE to baseForm.appendS())
        return WordForms().apply { put(wordForms, CHANGE_MODEL) }
    }

    private fun String.appendS() = when {
        endsWithVowel() -> substitute(0, "s")
        endsWithIon() -> substitute(2, "ones")
        endsWith("z") -> substitute(1, "ces")
        endsWithWeakConsonant() && !hasSetStress() -> setStress().substitute(0, "es")
        else -> substitute(0, "es")
    }

    private fun String.endsWithVowel() = endingIsInSet(VOWELS)
    private fun String.endsWithIon() = endsWith("ción") || endsWith("sión")
    private fun String.endsWithWeakConsonant() = endingIsInSet(WEAK_CONSONANTS)
    private fun String.hasSetStress() = STRESSED_VOWEL_PATTERN.matcher(this).matches()

    private fun String.setStress(): String {
        var count = 0
        for (i in this.indices.reversed()) {
            if (BASE_VOWELS.contains(this[i].toString())) count++
            if (count == 2) return substring(0, i) + this[i].setStress() + substring(i + 1)
        }
        return this
    }

    private fun Char.setStress() = when(this) {
        'e' -> 'é'
        'a' -> 'á'
        'o' -> 'ó'
        'i' -> 'í'
        'u' -> 'ú'
        else -> this
    }

    override fun getLanguageForms() = FORMS;
}
